File: examples/decode.ml
(* Decoding a fixed JSON data structure using Aeson.Decode *)
let mapJsonObjectString f decoder (encoder: int -> Js.Json.t) str =
  let json = Js.Json.parseExn str in
  Aeson.Decode.(dict decoder json)
    |> Js.Dict.map ((fun v -> f v) [@bs])
    |> Js.Dict.map ((fun v -> encoder v) [@bs])
    |> Aeson.Encode.dict
    |> Js.Json.stringify

let sum =
  Array.fold_left (+) 0

(* prints `{ "foo": 6, "bar": 24 }` *)
let _ =
  Js.log @@
    (mapJsonObjectString sum Aeson.Decode.(array int) Aeson.Encode.int {|
      {
        "foo": [1, 2, 3],
        "bar": [9, 8, 7]
      }
    |})

(* Error handling *)
let _ =
  let json = {|{ "y": 42 } |} |> Js.Json.parseExn in
  match Aeson.Decode.(field "x" int json) with
  | x ->
    Js.log x
  | exception Aeson.Decode.DecodeError msg ->
    Js.log ("Error:" ^ msg)

File: lib/js/examples/decode.js
'use strict';

var Js_dict = require("rescript/lib/js/js_dict.js");
var Core__Array = require("@rescript/core/lib/js/src/Core__Array.js");
var Aeson_decode = require("../src/Aeson_decode.js");
var Caml_js_exceptions = require("rescript/lib/js/caml_js_exceptions.js");

function mapJsonObjectString(f, decoder, encoder, str) {
  var json = JSON.parse(str);
  var m = Aeson_decode.dict(decoder, json);
  return JSON.stringify(Js_dict.map(encoder, Js_dict.map(f, m)));
}

function sum(xs) {
  return Core__Array.reduce(xs, 0, (function (prim0, prim1) {
                return prim0 + prim1 | 0;
              }));
}

console.log(mapJsonObjectString(sum, (function (param) {
            return Aeson_decode.array(Aeson_decode.$$int, param);
          }), (function (x) {
            return x;
          }), "\n      {\n        \"foo\": [1, 2, 3],\n        \"bar\": [9, 8, 7]\n      }\n    "));

var json = JSON.parse("{ \"y\": 42 } ");

var exit = 0;

var x;

try {
  x = Aeson_decode.field("x", Aeson_decode.$$int, json);
  exit = 1;
}
catch (raw_msg){
  var msg = Caml_js_exceptions.internalToOCamlException(raw_msg);
  if (msg.RE_EXN_ID === Aeson_decode.DecodeError) {
    console.log("Error:" + msg._1);
  } else {
    throw msg;
  }
}

if (exit === 1) {
  console.log(x);
}

exports.mapJsonObjectString = mapJsonObjectString;
exports.sum = sum;
/*  Not a pure module */
